### PR TITLE
chore: correct a bug in ActionBundle implementation

### PR DESCRIPTION
**Pull request by hmdebenque**
there was a problem with caching in action bundle. This commit solves this
by minimize the cache impact on code structure.
This cache is based on guava's implementation.

**Please check if the PR fulfills these requirements**
- [x] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] The code coverage on new code is > 75 % for backend and > 95% for frontend
- [x] The new code does not introduce new technical issues (sonar / eslint)
- [ ] Functional tests have been performed
- [ ] Docker configuration files for config-std or config-cloud profiles are impacted

**Please check the browsers you've tested on**
- [ ] Chrome, Firefox, Safari, Edge, IE11
- [ ] No, that's bad, this PR should not be merged !
- [ ] No, and no need to (backend changes only)

### CODE REVIEW

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/1152/) [:star:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/1152/)

File: dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java [hunk 1]
```diff
@@ -51,33 +77,20 @@
...
-        return clazz.getName() + "_" + locale.getLanguage();
-    }
-
-    private static String generateBundleKey(Class clazz) {
-        return ActionsBundle.generateBundleKey(clazz, null);
-    }
+    /**
+     * Simple cache.
+     */
+    // might be a good idea to make it customizable
```
Comment (fhuaulme) [inline dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: Just as Yoda said: just do it or don't :)

File: dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java [hunk 2]
```diff
@@ -205,17 +218,29 @@ private String formatMessage(final String template, final Object... args) {
...
+        String message;
+        String bundleMessage = getBundleValue(action, locale, code);
+        if (bundleMessage != null) {
+            message = formatMessage(bundleMessage, args);
+        } else {
+            message = null;
+        }
+        return message;
+    }
+
```
Comment (fhuaulme) [inline dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: Code format.

File: dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java [hunk 3]
```diff
@@ -225,49 +250,97 @@ private String getOptionalMessage(Object action, Locale locale, String code, Obj
...
-                try {
-                    bundle = ResourceBundle.getBundle(currentPackageName + '.' + ACTIONS_MESSAGES, locale);
+    /**
+     * Searches the list of bundles and returns the first that contains the key. <bold>warning:</bold> if the key is found
+     * in a bundle in a fallback language before being present in another bundle in the correct language the first will be
+     * returned.
+     */
+    private ResourceBundle findBundleContainingKey(Object action, Locale locale, String key, Collection<String> bundleFallbacks) {
+        ResourceBundle bundle = null;
+        ResourceBundle bundleSearched;
```
Comment (fhuaulme) [inline dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: Should be 'searchedBundle' and moreover, is it possible to have only one variable?

File: dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java [hunk 4]
```diff
@@ -140,7 +153,7 @@ public static String actionDocUrl(Object action, Locale locale, String actionNam
 
     /**
      * Fetches action label at {@code action.<action_name>.label} in the dataprep actions resource bundle. If message does not
-     * exist, code will lookup in {@link #fallBackKey} resource bundle (i.e. Data Prep one) for message.
+     * exist, code will lookup in fallBack resource bundles (i.e. Data Prep one) for message.
```
Comment (fhuaulme) [inline dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: "fallback" (not "fallBack").

Comment (fhuaulme) [reply dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: and "bundle" (not "bundles").

File: dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java [hunk 5]
```diff
@@ -205,17 +218,29 @@ private String formatMessage(final String template, final Object... args) {
...
-        final ResourceBundle bundle = findBundle(action, locale);
-        final String fallbackBundleKey = generateBundleKey(this.getClass());
-
-        // We can put some cache here if default internal caching it is not enough
-        if (Objects.nonNull(bundle) && bundle.containsKey(code)) {
-            return formatMessage(bundle.getString(code), args);
-        } else if (Objects.nonNull(actionToResourceBundle.get(fallbackBundleKey))
-                && actionToResourceBundle.get(fallbackBundleKey).containsKey(code)) {
-            return formatMessage(actionToResourceBundle.get(fallbackBundleKey).getString(code), args);
+        String message;
```
Comment (fhuaulme) [inline dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: final

File: dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java [hunk 6]
```diff
@@ -225,49 +250,97 @@ private String getOptionalMessage(Object action, Locale locale, String code, Obj
...
+    /**
+     * Create the list of bundle names fallback to search for i18n keys:
+     * <ul>
+     * <li> first add the full package hierarchy of the action class</li>
+     * <li> then add the full package hierarchy of this bundle class</li>
+     * <li> and last but not least, add the default {@link #BUNDLE_NAME}</li>
+     * </ul>
+     */
+    private Collection<String> getBundlesFallbackList(Object action) {
+        LinkedHashSet<String> packageHierarchy = new LinkedHashSet<>();;
```
Comment (fhuaulme) [inline dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: A set here? Could it cause an issue with a package such as "com.talend.com" ?

File: dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java [hunk 7]
```diff
@@ -225,49 +250,97 @@ private String getOptionalMessage(Object action, Locale locale, String code, Obj
...
         this.documentationUrlBase = documentationUrlBase;
     }
 
     private String getDocumentationUrlBase() {
         return documentationUrlBase;
     }
+
+    private static class CacheKey {
+
+        final Object action;
```
Comment (fhuaulme) [inline dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: private

File: dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java [hunk 8]
```diff
@@ -225,49 +250,97 @@ private String getOptionalMessage(Object action, Locale locale, String code, Obj
...
+        final Locale locale;
```
Comment (fhuaulme) [inline dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: + code format.

File: dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java (hunk 3 above)
Comment (hmdebenque) [reply dataprep-backend-common/src/main/java/org/talend/dataprep/i18n/ActionsBundle.java]: Possible but it would mean having 2 return statements

**Comment by build-talend-tdp** [comment]
Pull request test result: [success](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/1157/) [:star:](https://ci-tdp.datapwn.com/job/data-prep_pull-request_master/1157/)

**Pull request merged**